pkg/block/index.go

```diff
@@ -495,12 +495,16 @@ func Repair(logger log.Logger, dir string, id ulid.ULID, source metadata.SourceT
 	resmeta.Stats = tsdb.BlockStats{} // reset stats
 	resmeta.Thanos.Source = source    // update source
 
-	if err := rewrite(indexr, chunkr, indexw, chunkw, &resmeta, ignoreChkFns); err != nil {
+	if err := rewrite(logger, indexr, chunkr, indexw, chunkw, &resmeta, ignoreChkFns); err != nil {
 		return resid, errors.Wrap(err, "rewrite block")
 	}
 	if err := metadata.Write(logger, resdir, &resmeta); err != nil {
 		return resid, err
 	}
+	// TSDB may rewrite metadata in bdir
```

Comment:
The TSDB code handily rewrites the block metadata even though we don't write or change read block here.  This nukes the extra metadata Thanos uses.  Write it back.

Reply:
So this is also a bug that you are trying to fix? Perhaps we should add it to the changelog and/or the PR messsage.

Reply:
Not a bug, this is part of the initial patches to be able to reuse downloaded blocks.  The TSDB library isn't aware of the additional meta data Thanos provides and was removing it.  Therefore, when a previously downloaded block is reused it would have incomplete metadata without this change.

Reply:
So this is being changed now: https://github.com/prometheus/tsdb/pull/637 so this won't be needed TBH (once we use latest TSDB).

 Also this is true what @GiedriusS said - splitting PRs into focused changes helps a lot. However this change relies on this fix. I would be more explicit in description. This will help us to understand this early on (:

Reply:
Let's add TODO to update TSDB and remove this as it will be not needed in newest TSDB. For now we need this, thanks for fixing this :+1:

Reply:
Todo added, thanks!